### PR TITLE
New lint to detect `&std::path::MAIN_SEPARATOR.to_string()`

### LINKED ISSUE #10480: Suggest `std::path::MAIN_SEPARATOR_STR`
### What it does

Suggests `std::path::MAIN_SEPARATOR_STR` over `std::path::MAIN_SEPARATOR.to_string()`

### Lint Name

_No response_

### Category

_No response_

### Advantage

_No response_

### Drawbacks

_No response_

### Example

```rust
&std::path::MAIN_SEPARATOR.to_string()
```

Could be written as:

```rust
std::path::MAIN_SEPARATOR_STR
```

### PR DESCRIPTION
**Pull request by samueltardieu**
Fixes #10480 

changelog: [`manual_main_separator_str`] new lint

### CODE REVIEW

**Comment by llogiq** [comment]
I wonder if we could extend the `unnecessary_to_owned` lint instead of creating a new one. Also doesn't that work with all `&str` constants? Why single out `MAIN_SEPARATOR`?

**Comment by samueltardieu** [comment]
> I wonder if we could extend the `unnecessary_to_owned` lint instead of creating a new one. Also doesn't that work with all `&str` constants? Why single out `MAIN_SEPARATOR`?

It depends on the existence of a `&str` constant (`std::path::MAIN_SEPARATOR_STR`) with the same content as the `char` constant (`std::path::MAIN_SEPARATOR`).

File: tests/ui/manual_main_separator_str.fixed [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// run-rustfix
+
+#![allow(unused)]
+#![warn(clippy::manual_main_separator_str)]
+
+use std::path::MAIN_SEPARATOR;
+
+fn main() {
```
Comment (llogiq) [inline tests/ui/manual_main_separator_str.fixed:8]: What about `let`s without a given type? Or other expressions that are not direct local bindings?

**Review by llogiq — COMMENTED** [review]
Ah, I hadn't noticed the difference here.

I'd like to see some more tests, otherwise this looks good.

File: tests/ui/manual_main_separator_str.fixed (hunk 1 above)
Comment (samueltardieu) [reply tests/ui/manual_main_separator_str.fixed:21]: I have added new tests, such as initializing a `struct` field with this expression or passing it as a function argument.

I've also made it work when it is a method receiver (see the test with `encode_utf16()`), without an explicit `&`.

**Comment by samueltardieu** [comment]
@llogiq Is there anything more you'd like me to do on this one?

**Pull request merged**